Code generation optimizations

martinus opened the pull request:
After #9266 I did some profiling of doxygen generation of our C++ codebase. Based on that I have done a few optimizations that bring a 9x speedup, from ~49 minutes down to just over 5 minutes.

The most important change was the first commit, which using `const& for LinkedRefMap` instead of copies.

Note that the commit "Used hash data structures for SymbolMap" seems to cause the regression test `027_extends.c` to fail, not sure if this is an issue?

albert-github commented:
The:
> Note that the commit "Used hash data structures for SymbolMap" seems to cause the regression test 027_extends.c to fail, not sure if this is an issue?

is an issue, either the test results so far were incorrect (means update of test result required) or are now incorrect (code has to be corrected).
To visually check the test results you can run `make tests TEST_FLAGS="--keep --xml --xhtml --id 27"` and look in the `testing/test_output_027` subdirectory of your build environment.

I did the "approve and run" on the GitHub Actions workflow to see what kind of errors are given by the `027_extends.c` test, without looking into your code.

martinus commented:
Thanks @doxygen! Yes, I used [hotspot](https://github.com/KDAB/hotspot) which shows a nice flamegraph of the hotspots. It's awesome, I use it a lot to find bottlenecks. I saw lots of copies happen to data that's only read, so I tried to use `const&` when possible and it seems to have worked very well :slightly_smiling_face: 

I'm building doxygen with RelWithDebInfo like so:

```sh
CC="ccache clang" CXX="ccache clang++" CMAKE_BUILD_TYPE=RelWithDebInfo cmake -G Ninja ../src
```

Then hotspot is able to show nice call stacks, and lots of copies of `std::unordered_map` happen which are really costly.

I've rebased and removed the problematic change, now the test work for me
